contrib/fmhash/fmhash.c

```diff
@@ -0,0 +1,385 @@
+			, (func->nParams == 2 ? &seedVal : NULL)
+			, hcontext)));
+	ret->d.n = hcontext->xhash;
+finalize_it:
+	varFreeMembers(&hashStrVal);
+	if(func->nParams == 2) varFreeMembers(&seedVal);
+}
+
+static void ATTR_NONNULL()
+fmHashXXmod(struct cnffunc *__restrict__ const func, struct svar *__restrict__ const ret,
+		void *__restrict__ const usrptr, wti_t *__restrict__ const pWti) {
+
+	DEFiRet;
+	struct svar hashStrVal;
+	struct svar modVal;
+	struct svar seedVal;
+	hash_context_t* hcontext = NULL;
+	cnfexprEval(func->expr[0], &hashStrVal, usrptr, pWti);
+	cnfexprEval(func->expr[1], &modVal, usrptr, pWti);
+	if(func->nParams == 3) cnfexprEval(func->expr[2], &seedVal, usrptr, pWti);
+	ret->d.n = 0;
+	ret->datatype = 'N';
+	hcontext = (hash_context_t*) func->funcdata;
+	CHKiRet((hcontext->hash_wrapper_2_3(&hashStrVal
+			, &modVal, func->nParams > 2 ? &seedVal : NULL
+			, hcontext)));
+	ret->d.n = hcontext->xhash;
+finalize_it:
+	varFreeMembers(&hashStrVal);
+	varFreeMembers(&modVal);
+	if(func->nParams == 3) varFreeMembers(&seedVal);
+}
+
+static rsRetVal ATTR_NONNULL(1)
+init_fmHash64(struct cnffunc *const func)
+{
+	DEFiRet;
+	hash_context_t *hash_context = NULL;
+	if(func->nParams < 1) {
+		parser_errmsg("fmhash: hash64(string) / hash64(string, seed)"
+				" insufficient params. %d of file %s\n",
+			__LINE__, __FILE__);
+		iRet = RS_RET_ERR;
+		FINALIZE;
+	}
+	func->destructable_funcdata = 1;
+	CHKmalloc(hash_context = calloc(1, sizeof(hash_context_t)));
+	init_hash64_context(hash_context);
+	func->funcdata = (void*)hash_context;
+
+finalize_it:
+	RETiRet;
+}
+
+static rsRetVal ATTR_NONNULL(1)
+init_fmHash64mod(struct cnffunc *const func)
+{
+	DEFiRet;
+	hash_context_t *hash_context = NULL;
+	if(func->nParams < 2) {
+		parser_errmsg("fmhash: hash64mod(string, mod)/hash64mod(string, mod, seed)"
+				" insufficient params. %d of file %s\n",
+			__LINE__, __FILE__);
+		iRet = RS_RET_ERR;
+		FINALIZE;
+	}
+	func->destructable_funcdata = 1;
+	CHKmalloc(hash_context = calloc(1, sizeof(hash_context_t)));
+	init_hash64_context(hash_context);
+	func->funcdata = (void*)hash_context;
+finalize_it:
+	RETiRet;
+}
+
+static rsRetVal ATTR_NONNULL(1)
+init_fmHash32(struct cnffunc *const func)
+{
+	DEFiRet;
+	hash_context_t *hash_context = NULL;
+	if(func->nParams < 1) {
+		parser_errmsg("fmhash: hash32(string) / hash32(string, seed)"
+				" insufficient params. %d of file %s\n",
+			__LINE__, __FILE__);
+		iRet = RS_RET_ERR;
+		FINALIZE;
+	}
+	func->destructable_funcdata = 1;
+	CHKmalloc(hash_context = calloc(1, sizeof(hash_context_t)));
+	init_hash32_context(hash_context);
+	func->funcdata = (void*)hash_context;
+
+finalize_it:
+	RETiRet;
+}
+
+static rsRetVal ATTR_NONNULL(1)
+init_fmHash32mod(struct cnffunc *const func)
+{
+	DEFiRet;
+	hash_context_t *hash_context = NULL;
+	if(func->nParams < 2) {
+		parser_errmsg("fmhash: hash32mod(string, mod)/hash32mod(string, mod, seed)"
+				" insufficient params. %d of file %s\n",
```

Comment:
source code locations should only be given for internal errors -- else it tends to confuse users, which then often think it **is** an internal error (I made this too often myself...).